### DISCUSSION TITLE
How to trace Pydantic serializer warning without traceback

### DISCUSSION THREAD

**Discussion by jamesbraza** [pydantic/pydantic#9317] [Question]
In my application logs this warning shows up:

```none
/usr/local/lib/python3.12/site-packages/pydantic/main.py:347: UserWarning: Pydantic serializer warnings:
  Expected `int` but got `str` - serialized value may not be as expected
  return self.__pydantic_serializer__.to_python(
```

The traceback doesn't seem to propagate further than this, so I have no idea where in my code this error is coming from.

How can I hunt this down further?

**Comment by jbonilla-tao** [comment]
bump

**Comment by Viicos** [comment]
This warning is emitted in `pydantic-core`, so there's no traceback available (as it isn't Python code). However, I think it might be possible to give more info in the message

**Reply by jamesbraza** [reply]
Hello @Viicos happy to see your name here.

So I inherited a sizable app that is also nondeterministic in nature, and this message seems to randomly appears in the logs. I have failed to figure out where it comes from, hence why I made the original post here. Any tips on how to figure out where the error comes from?

---

Thinking about this, Pydantic internals emitting a warning without context is not ideal. Is there any way Pydantic can link a full stack trace, or use f-strings to show the field name having a type mismatch?

**Comment by jbonilla-tao** [comment]
I find myself in a very similar position as well. An improvement to the logs would be greatly appreciated!

**Comment by aodj-kara** [comment]
Logging the class name would be great to have to help debug these warnings.

**Comment by sneakers-the-rat** [comment]
are there any rust ppl that can show a poor soul how to set up a debug environment for this? i will repay in trying to add more information to the warnings here

**Comment by mohit2152sharma** [comment]
Tagging along, as I am also in the same boat. (Encountering a similar issue)

**Comment by zomglings** [comment]
In the same boat.

Echoing @aodj-kara :
Even though the stack trace is not available since the warning is not coming from Python, it would be super helpful if the message showed which Pydantic class was raising the warning.

Edit: Solved my particular issue by stepping through with the debugger. Wasn't a great experience. In addition to the class name, I would also like to know the field which is raising the warning.

**Reply by arvindsaripalli** [reply]
+1 for a very nested pydantic object this was a pain to figure out, but with debugger I was able to find which field was the problem.

**Comment by Viicos** [comment]
https://github.com/pydantic/pydantic-core/pull/1799 Adds the field name as a start to giving better location context. Next steps tracked in https://github.com/pydantic/pydantic/issues/10495.
